Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm

```diff
@@ -1044,6 +1044,76 @@
 
 #endif
 
+static bool elementHasHiddenVisibility(StyledElement* styledElement)
+{
+    auto* inlineStyle = styledElement->inlineStyle();
+    if (!inlineStyle)
+        return false;
+
+    RefPtr value = inlineStyle->getPropertyCSSValue(CSSPropertyVisibility);
+    if (!value)
+        return false;
+
+    return value->valueID() == CSSValueHidden;
+}
+
+void WebPage::createTextIndicatorForElementWithID(const String& elementID, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>&&)>&& completionHandler)
+{
+    RefPtr frame = m_page->checkedFocusController()->focusedOrMainFrame();
+    if (!frame) {
+        ASSERT_NOT_REACHED();
+        completionHandler(std::nullopt);
+        return;
+    }
+
+    RefPtr document = frame->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        completionHandler(std::nullopt);
+        return;
+    }
+
+    RefPtr element = document->getElementById(elementID);
+    if (!element) {
+        ASSERT_NOT_REACHED();
+        completionHandler(std::nullopt);
+        return;
+    }
+
+    RefPtr styledElement = dynamicDowncast<StyledElement>(element.get());
+    if (!styledElement) {
+        ASSERT_NOT_REACHED();
+        completionHandler(std::nullopt);
+        return;
+    }
+
+    // Temporarily force the content to be visible so that it can be snapshotted.
+
+    auto initialVisibility = elementHasHiddenVisibility(styledElement.get()) ? CSSValueHidden : CSSValueVisible;
+
+    styledElement->setInlineStyleProperty(CSSPropertyVisibility, CSSValueVisible, IsImportant::Yes);
+
+    auto elementRange = WebCore::makeRangeSelectingNodeContents(*styledElement);
+
+    std::optional<WebCore::TextIndicatorData> textIndicatorData;
+    constexpr OptionSet textIndicatorOptions {
+        WebCore::TextIndicatorOption::IncludeSnapshotOfAllVisibleContentWithoutSelection,
+        WebCore::TextIndicatorOption::ExpandClipBeyondVisibleRect,
+        WebCore::TextIndicatorOption::SkipReplacedContent,
+        WebCore::TextIndicatorOption::RespectTextColor
+    };
+
+    RefPtr textIndicator = WebCore::TextIndicator::createWithRange(elementRange, textIndicatorOptions, WebCore::TextIndicatorPresentationTransition::None, { });
+    if (!textIndicator) {
+        completionHandler(std::nullopt);
+        return;
+    }
+
+    styledElement->setInlineStyleProperty(CSSPropertyVisibility, initialVisibility, IsImportant::Yes);
```

Comment:
This isn't really idempotent, because you could be adding inline style to an element that never had one.

Reply:
what should I do instead?

Reply:
`removeInlineStyleProperty` if it never had one?